components/tidb_query_expr/src/impl_string.rs

```diff
@@ -1014,6 +1014,52 @@ pub fn repeat(input: BytesRef, cnt: &Int, writer: BytesWriter) -> Result<BytesGu
     Ok(writer.finish())
 }
 
+#[rpn_fn(writer)]
+#[inline]
+pub fn substring_2_args_utf8(
+    input: BytesRef,
+    pos: &Int,
+    writer: BytesWriter,
+) -> Result<BytesGuard> {
+    substring_utf8(input, *pos, input.chars().count() as Int, writer)
+}
+
+#[rpn_fn(writer)]
+#[inline]
+pub fn substring_3_args_utf8(
+    input: BytesRef,
+    pos: &Int,
+    len: &Int,
+    writer: BytesWriter,
+) -> Result<BytesGuard> {
+    substring_utf8(input, *pos, *len, writer)
+}
+
+#[inline]
+fn substring_utf8(input: BytesRef, pos: Int, len: Int, writer: BytesWriter) -> Result<BytesGuard> {
+    let (len, len_positive) = i64_to_usize(len, len > 0);
+    let (pos, positive_search) = i64_to_usize(pos, pos > 0);
+    if pos == 0 || len == 0 || !len_positive {
+        return Ok(writer.write_ref(Some(b"")));
+    }
+
+    let char_len = input.chars().count();
+    let start = if positive_search {
+        (pos - 1).min(char_len)
+    } else {
+        char_len.checked_sub(pos).unwrap_or(char_len)
+    };
+    let input = str::from_utf8(input)?;
+    Ok(writer.write(Some(
+        input
+            .chars()
+            .skip(start)
+            .take(len)
+            .collect::<String>()
+            .into_bytes(),
```

Comment:
Could you try to avoid copying the substring?

Reply:
done